sandbox/libs/authn/src/main/java/org/opensearch/authn/Subject.java

```diff
@@ -30,4 +31,10 @@ public interface Subject {
      */
     public void login(final AuthenticationToken token);
 
+    /**
+     * The subject will confirm if it has the permissions that are checked against, if it has those permissions it returns null, otherwise it returns the exception.
+     * Note; this method not throw the exception
+     * */
+    public UnauthorizedException checkPermission(List<String> permissions);
```

Comment:
I see that `shiro` has a similar check, but throws the exception and makes the method void.

```
void checkPermission(String permission) throws AuthorizationException;
```

Curious, what do you see as the advantage to this as opposed to throwing an exception like shiro? From what I can see in this PR, it looks like we may have to do `null` checks after each check of permissions. If an `UnauthorizedException` is thrown we may be able to handle the exception centrally.

Reply:
There is an established pattern [1] in TransportAction.java to check if an exception is returned by `validate(...)`.  While there could be an issue with an exception result being ignored, at least in this use case `throw` would cause a problem since the exception needs to be captured and then listener be sent a failure.

While we could provide two methods something like `throwIfNoPermission(...)` / `checkPermission(...)`, this might create more confusion compared to a singular model.  What do you think?

[1] https://github.com/opensearch-project/OpenSearch/blob/014adb4a1b0133538441f51a55599a0baadfc0b8/server/src/main/java/org/opensearch/action/support/TransportAction.java#L182-L186

Reply:
Thanks for the context! I didn't realize this was an already established pattern.

Reply:
oo.. I didn't know either. Thanks Peter.